microsoft/testsuites/dpdk/dpdkutil.py

```diff
@@ -610,3 +619,455 @@ def _parallel_cleanup(node: Node) -> None:
+    # generate the dpdk include arguments to add to our commandline
+    include_devices = [
+        fwd_kit.testpmd.generate_testpmd_include(
+            subnet_a_nics[forwarder], dpdk_port_a, force_netvsc=True
+        ),
+        fwd_kit.testpmd.generate_testpmd_include(
+            subnet_b_nics[forwarder], dpdk_port_b, force_netvsc=True
+        ),
+    ]
+
+    # Generating port,queue,core mappings for forwarder
+    # NOTE: For DPDK 'N queues' means N queues * N PORTS
+    # Each port P has N queues (really queue pairs for tx/rx)
+    # Queue N for Port A and Port B will be assigned to the same core.
+    # l3fwd rquires us to explicitly map these as a set of tuples.
+    # Create a set of tuples (PortID,QueueID,CoreID)
+    # These are minimally error checked, you can accidentally assign
+    # cores and queues to unused ports etc. and only get runtime spew.
+    queue_count = get_l3fwd_queue_count(
+        available_cores,
+        force_single_queue=force_single_queue,
+        is_mana=fwd_kit.testpmd.is_mana,
+    )
+
+    config_tups = []
+    included_cores = []
+    last_core = 1
+    # create the list of tuples for p,q,c
+    # 2 ports, N queues, N cores for MANA
+    # 2 ports, N queues, 2N cores for MLX
+    for q in range(queue_count):
+        config_tups.append((dpdk_port_a, q, last_core))
+
+        if not fwd_kit.testpmd.is_mana:
+            included_cores.append(str(last_core))
+            last_core += 1
+        config_tups.append((dpdk_port_b, q, last_core))
+        # add the core ID to our list of cores to include
+        included_cores.append(str(last_core))
+        last_core += 1
+
+    # pick promiscuous mode arg, note mana doesn't support promiscuous mode
+    if fwd_kit.testpmd.is_mana:
+        promiscuous = ""
+    else:
+        promiscuous = "-P"
+
+    # join all our options into strings for use in the commmand
+    joined_configs = ",".join([f"({p},{q},{c})" for (p, q, c) in config_tups])
+    joined_include = " ".join(include_devices)
+    # prefer the '-l 1,2,3' arg version over '-l 1-4' form to avoid a dpdk bug
+    joined_core_list = ",".join(included_cores)
+    fwd_cmd = (
+        f"{server_app_path} {joined_include} -l {joined_core_list}  -- "
+        f" {promiscuous} -p {get_dpdk_portmask([dpdk_port_a,dpdk_port_b])} "
+        f' --lookup=lpm --config="{joined_configs}" '
+        "--rule_ipv4=rules_v4  --rule_ipv6=rules_v6 --mode=poll --parse-ptype"
+    )
+    # START THE TEST
+    # finally, start the forwarder
+    fwd_proc = forwarder.execute_async(
+        fwd_cmd,
+        sudo=True,
+        shell=True,
+    )
+    try:
+        fwd_proc.wait_output("L3FWD: entering main loop", timeout=30)
+    except LisaException:
+        raise LisaException(
+            "L3fwd did not start. Check command output for incorrect flags, "
+            "core dumps, or other setup/init issues."
+        )
+
+    # start ntttcp client and server
+    ntttcp_threads_count = 64
```

Comment:
is it expected that fwd_proc running in the background during run ntttcp? if so, I saw that ntttcp starts to run after fwd_proc finishes running

Reply:
How so? Wait_output doesn't kill the process, it just checks stdout. The flow is:
```
fwd.start_async() # no timeout
fwd.wait_output('started', timeout=30)  # blocks, throw if timeout
server.start_async(timeout=30) 
client.start(timeout=10) # blocks
server.wait_result()
fwd.kill()
```

Reply:
This might be confusing because l3fwd doesn't log much while it's running? You can verify the process is up and killed after the other two exit though. Packets won't flow between the subnets without l3fwd running.